### ISSUE TITLE
Support Blob POCO bindings

### ISSUE THREAD

**Issue by mathewc** [Azure/azure-webjobs-sdk#995]
Remove ICloudBlobStreamBinder (replace with custom Converters). The blob binding had an existing extensibility mechanism [ICloudBlobStreamBinder](https://github.com/Azure/azure-webjobs-sdk/blob/663a508e8a851629c26a51e7de3af36629dfd120/src/Microsoft.Azure.WebJobs/ICloudBlobStreamBinder.cs) that customers can use to register custom blob stream => POCO types.

This also causes problems for people because they expect to be able to **bind blobs to POCO types directly** as they can with all other bindings. E.g. see SO question [here](https://github.com/Azure/azure-webjobs-sdk/issues/995).

When the blob binding is converted over to the new binding APIs, we can remove that old extensibility mechanism.

**Comment by fabiocav** [comment]
relates to #1001

**Comment by brettsam** [comment]
I'm pretty sure this isn't actually fixed. In moving to our new DI patterns, I removed the legacy ICloudBlobStreamBinder registration code and a bunch of tests started failing. I also don't see anywhere in the blob binding setup that indicates we support binding to POCOs. I'm going to re-enable this until we can take a deeper look and figure out what we need to do.

Also note that we have official docs on this that we'll need to update when we remove this interface: https://docs.microsoft.com/en-us/azure/visual-studio/vs-storage-webjobs-getting-started-blobs#getting-serialized-blob-content-by-using-icloudblobstreambinder

**Comment by brettsam** [comment]
Note: Some tests pass due to this binder: https://github.com/Azure/azure-webjobs-sdk/blob/dev/test/Microsoft.Azure.WebJobs.Host.EndToEndTests/BlobToCustomObjectBinder.cs. It's likely that logic needs to move into the core binding wire-up.

Part of closing this issue should be removal of that interface -- assuming all scenarios can be covered without it.

**Comment by fabiocav** [comment]
Running into issues because of this work item again. There's logic in script that explicitly calls the methods to support the legacy behavior, which, unfortunately, doesn't play well with the assembly isolation work we've put in place.

Let's confirm that we have functionality coverage and remove the legacy code (which should no longer be needed)

**Comment by brettsam** [comment]
When doing the DI tests, I confirmed that we do *not* have support for Blob POCOs. We'll need to add it.

**Comment by mathewc** [comment]
Right - we don't have POCO support. The old ICloudStreamBinder stuff was how you did POCOs. That's what this issue was created for.

**Comment by mathewc** [comment]
We also have a test here https://github.com/Azure/azure-webjobs-sdk/blob/254a9b6cdd07313f695941349739b9fa22c6a888/test/Microsoft.Azure.WebJobs.Host.FunctionalTests/HostCallTests.cs#L777 that relies on the current shim layer for this: https://github.com/Azure/azure-webjobs-sdk/blob/254a9b6cdd07313f695941349739b9fa22c6a888/test/Microsoft.Azure.WebJobs.Host.FunctionalTests/HostCallTests.cs#L777.

**Comment by thomaslevesque** [comment]
The [wiki](https://github.com/Azure/azure-webjobs-sdk/wiki/Blobs) still mentions `ICloudStreamBinder`, which is no longer there. I can't find any information on how to consume a POCO directly as a webjob function parameter. Is it even possible?

**Comment by mathewc** [comment]
Yes, you can register your own converter to do this via an extension. Say you have a WorkItem POCO. First write an IExtensionConfigProvider that registers converters as needed:

```csharp
internal class CustomBlobConverterExtensionConfigProvider : IExtensionConfigProvider
{
    public void Initialize(ExtensionConfigContext context)
    {
        context.AddConverter<Stream, WorkItem>(s =>
        {
            // read and convert
            return new WorkItem();
        });

        context.AddConverter<ApplyConversion<WorkItem, Stream>, object>(p =>
        {
            WorkItem value = p.Value;
            var stream = p.Existing;

            TextWriter writer = new StreamWriter(stream);
            // write the value to the stream
            writer.FlushAsync().GetAwaiter().GetResult();

            return null;
        });
    }
}
```

Then, you register this config provider with the host builder on startup:

```csharp
var builder = new HostBuilder()
    .ConfigureWebJobs(b =>
    {
        b.AddAzureStorageCoreServices()
        .AddAzureStorage()
        .AddExtension<CustomBlobConverterExtensionConfigProvider>();
    })
```

Then you can bind your BlobTrigger function directly to your POCO:

```csharp
public void BlobTrigger([BlobTrigger("test")] WorkItem item, ILogger logger)
{
    // process item
}
```

**Comment by thomaslevesque** [comment]
Thanks @mathewc!
